geps/gep-1651.md

```diff
@@ -0,0 +1,298 @@
+Gateways using the same GatewayClass on the cluster, but ensure that the
+"cluster" service is only routable within the cluster. This would greatly
+simplify deployment for users over the instructions we have today.
+
+Likewise another use case is to provide load balancing capabilities within a virtual
+private network. Different IaaS providers offer private load balancers to support
+these use cases.
+
+## API
+
+We propose adding a new `routability` field under the `spec.infrastructure` stanza of a Gateway.
+
+### Predefined Routability Values
+
+Implementations MAY implement the following values for 'routability' and MUST abide by
+their defined semantics.
+
+Value | Scope
+-|-
+`Public`|The address is routable on the public internet
+`Private`|The address is routable inside a private network larger than a single cluster (ie. VPC) and MAY include RFC1918 address space
+`Cluster`|The address is routable inside the [cluster's network](https://kubernetes.io/docs/concepts/cluster-administration/networking/#how-to-implement-the-kubernetes-network-model)
+
+Values can be compared semantically - `Public` has a larger scope than `Private`, while `Private` has a larger scope than `Cluster`.
+
+### Vendor prefixed values
+
+Implementations can define custom 'routability' values by specifying a vendor prefix followed
+by a slash `/` and a custom name ie. `com.example.com/my-routability`.
+
+Comparing vendor prefixed scopes with the pre-defined ones in implementation specific.
+
+### Default Routability
+
+The default value of `routability` is implementation specific. It is RECOMMENDED that
+the default `routability` remains consistent for Gateways with the same
+`gatewayClassName`.
+
+Implementations MAY leave the `spec.infrastructure.routability` property unset and instead
+signal the default routability using the Gateway's `status.addresses`. See 'Status Addresses`
+for more details.
+
+### Mutability
+Implementations MAY prevent end-users from updating the `routability` value of a Gateway. If
+updates are allowed the semantics and behaviour will depend on the underlying implementation.
+
+### Go
+
+```go
+
+type GatewayRoutability string
+
+const (
+  GatewayRoutabilityPublic  GatewayRoutability  = "Public"
+  GatewayRoutabilityPrivate GatewayRoutability  = "Private"
+  GatewayRoutabilityCluster GatewayRoutability  = "Cluster"
+)
+
+type GatewaySpec struct {
+  // Infrastructure defines infrastructure level attributes about this Gateway instance.
+  Infrastructure GatewayInfrastructure `json:"infrastructure"`
+  // ...
+}
+type GatewayInfrastructure struct {
+  // AddressScope allows the Gateway to specify the accessibility of it's addresseses
+  Routability GatewayRoutability `json:"routability"`
+}
+
+type GatewayAddress struct {
+  // Routability specifies the routable bounds of this address
+  // Predefined values are - Private, Public, Cluster
+  //
+  // This only appears on the addresses in the Gateway's status
```

Comment:
I agree with this line, but the field will need to be optional and ignored in spec, and should have a doc note explaining that using it in spec will be ignored as well.

Reply:
Good catch. I think we really need to have separate types for addresses in spec and status to avoid any potential confusion here.

Reply:
Added a separate type